ksmt-core/src/main/kotlin/org/ksmt/expr/rewrite/KExprSubstitutor.kt

```diff
@@ -43,12 +43,11 @@ open class KExprSubstitutor(ctx: KContext) : KNonRecursiveTransformer(ctx) {
     override fun <T : KSort> transformExpr(expr: KExpr<T>): KExpr<T> =
         exprExprSubstitution[expr]?.uncheckedCast() ?: expr
 
-    override fun <T : KSort, A : KSort> transformApp(expr: KApp<T, A>): KExpr<T> =
-        transformAppAfterArgsTransformed(expr) { transformedArgs ->
-            val decl = declDeclSubstitution[expr.decl]?.uncheckedCast() ?: expr.decl
-            val transformedApp = decl.apply(transformedArgs)
-            return transformExpr(transformedApp)
-        }
+    override fun <T : KSort, A : KSort> transformApp(expr: KApp<T, A>): KExpr<T> {
+        val substitution: KDecl<T> = declDeclSubstitution[expr.decl]?.uncheckedCast() ?: return transformExpr(expr)
```

Comment:
Usually, `cast` is not applicable because of reified type parameters. I prefer to use `uncheckedCast`  to avoid  using two different functions everywhere.